### PR TITLE
Github action for custom metric tests with WPT API

### LINKED ISSUE #6: Integrate WPT GitHub action
https://github.com/WPO-Foundation/webpagetest-github-action

When someone creates a PR to change a custom metric, the GH action should automatically test it on WPT and comment on the PR with the output of the affected custom metric(s) and a link to the results.

It would also be nice to have a way to tell the GH action which URL(s) to use for testing and whether to use desktop and/or mobile, otherwise there should be sensible defaults. For example, the action could search the PR description for URLs and test config options like desktop/mobile.

### PR DESCRIPTION
**Pull request by max-ostapenko**
Resolves #6 

As pointed in [README](https://github.com/HTTPArchive/custom-metrics#testing), each custom metric adjustment can be tested with WPT.

PR includes:
- [x] test workflow,
- [x] WPT runner script,
- [x] unit tests,
- [x] parsing custom websites to test from PR body (see PR template).

---
**Test websites**:

- https://example.com/

### CODE REVIEW

**Comment by rviscomi** [comment]
Oh wow thank you for picking this up!

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [inline .github/workflows/wpt-test.yml:36]: @rviscomi could you add a WPT secret as expected here

Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:36]: And if you could share one to speed up local dev - would be great!

Comment (rviscomi) [reply .github/workflows/wpt-test.yml:36]: Set up the `WPT_API_KEY` repo secret and DM'd you in the HTTP Archive slack with a private key you can use for testing.

**Comment by max-ostapenko** [comment]
Unfortunately WebPageTest/webpagetest-api doesn't support POST requests to WPT API.
So when we try to attach custom metrics code (privacy.js ~ 14Kb) - test run request fails.

I don't expect [WPT issue](https://github.com/WebPageTest/webpagetest-api/issues/177) to be resolved soon.
So I think of launching first version with limited size support only. And later introduce support of POST requests.

@rviscomi @pmeenan do you have alternative ideas/comments?

**Comment by pmeenan** [comment]
Could also post to the API directly. Though it's more work to implement the
polling. Or fork the project and implement post (might be easier).

**Comment by max-ostapenko** [comment]
Thanks @pmeenan
I think I see the easy way to make a fork and probably even make a PR to WPT.

**Comment by max-ostapenko** [comment]
I tested API as follows:
```bash
curl --location 'https://www.webpagetest.org/runtest.php?url=https%3A%2F%2Fexample.com%2F&f=json&k=INSERT_WPT_KEY' \
--header 'Content-Type: application/json' \
--data '{
    "location": "Dulles:Chrome.Native",
    "mobile": 1
}'
```
but still got `mobile: 0` and `connectivity: Cable` in the test results.

@pmeenan Did I misunderstood the doc and [runtest.php](https://www.webpagetest.org/runtest.php) doesn't look at parameters passed in POST request body, just URL query?
OR am I doing something wrong?

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [inline .github/workflows/wpt-test.yml:27]: GET request URL accepted by WPT server is around 6Kb long.
We can start with limited support and see how helpful the tests are.

**Comment by max-ostapenko** [comment]
@pmeenan Maybe there is some option to use our custom WPT instance to run the tests with complete custom metrics code included?
I mean it works somehow in PROD crawling.

Otherwise, @rviscomi @pmeenan, looking forward to your review.

**Comment by rviscomi** [comment]
@pmeenan WDYT about using the private instance for testing? My only concern would be somehow interfering with the crawl queue.

**Comment by pmeenan** [comment]
Sorry for the slow replies. Came down with covid after IETF :(

The private instance has a single test machine and is independent of the
crawl. I'd just need to assign a key.

As far as the API goes, POST should work as long as it is multipart mime.
And not mixed with query params and post fields.

**Comment by max-ostapenko** [comment]
[WPT API library PR](https://github.com/WebPageTest/webpagetest-api/pull/178) took quite some time, but I hope it will be merged into original wrapper.

For now we'll use a forked version.

**Comment by max-ostapenko** [comment]
Now you could even include all custom metrics in a single test case.
But by default - only chapter related to keep it fast.

File: tests/wpt.js
Comment (rviscomi) [inline tests/wpt.js:5]:
```suggestion
const wptApiKey = process.env.WPT_API_KEY;
```

File: tests/wpt.js
Comment (rviscomi) [inline tests/wpt.js:29]: The response could be huge, not sure if it's helpful to log the whole thing. Maybe just the test ID:

```suggestion
        console.log(`Test ID: ${JSON.parse(response.data.id)}`);
```

File: tests/wpt.js
Comment (rviscomi) [inline tests/wpt.js:6]:
```suggestion
const wpt = new WebPageTest(wptServer, wptApiKey);
```

**Review by rviscomi — CHANGES_REQUESTED** [review]

File: dist/privacy.test.js
Comment (max-ostapenko) [reply dist/privacy.test.js:1]: Moved.
Where is the script to integrate metrics for the prod tests?
I think there are quite some things that could be reused.

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:1]: These are available in action logs.
My idea was that it's better to keep PR clean from test logs, because it not helpful for review discussion.

File: tests/wpt.js
Comment (max-ostapenko) [reply tests/wpt.js:29]: I kept the test results URL.

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:23]: Oh, I think I misunderstood the files.
Now I see that each file is integrated as a custom metrics (has a return statement in it).

I removed the list. If there are any exceptions - let me know.

File: .github/workflows/wpt-test.yml
Comment (rviscomi) [reply .github/workflows/wpt-test.yml:1]: We used to manually put test results in the PR comments ([example](https://github.com/HTTPArchive/custom-metrics/pull/24)) as a way of helping reviewers confirm that the changes are working working and see what the expected output is. For me at least, I think it's helpful. But let me know if it's too much work.

File: dist/privacy.test.js
Comment (rviscomi) [reply dist/privacy.test.js:1]: Somewhere in the core WebPageTest code, maybe @pmeenan can point you to it.

File: tests/privacy.test.js
Comment (rviscomi) [inline tests/privacy.test.js:1]: Now that these are in the `tests/` directory, do we need the `.test.js` suffix or can we simplify to `.js`?

File: tests/privacy.test.js
Comment (rviscomi) [inline tests/privacy.test.js:11]: Having so many external dependencies makes this test more likely to fail if any of them change. I wonder if we should have more contrived test cases on pages that we control to prevent that?

Also curious to get @tunetheweb's opinion.

Comment (max-ostapenko) [reply tests/privacy.test.js:11]: I've tried to do testing previously with sites like:
- https://global-privacy-control.glitch.me/
- https://user-agent-client-hints.glitch.me/
but it doesn't cover all the cases even for privacy chapter.

That would be awesome to have some more extended library of browser API and other demos.
Then we would be able to improve testing speed and maybe run all the test-cases with each action run.

File: tests/privacy.test.js
Comment (max-ostapenko) [reply tests/privacy.test.js:1]: It's a default behaviour with Jest (cares about suffix but not location).
And allowed to differentiate between tests and other files (but I'd rather move wpt API module to different location - `utils/wpt.js`?)

To get rid of `.test` part I could customise npm action: `"test": "jest --testMatch=**/tests/**/*.js"`
What do you think?

Comment (rviscomi) [reply tests/privacy.test.js:1]: Oh I guess it's ok if it's a semantic thing with Jest.

File: .github/workflows/wpt-test.yml
Comment (rviscomi) [reply .github/workflows/wpt-test.yml:1]: Cool to see this working! https://github.com/HTTPArchive/custom-metrics/actions/runs/5814745249/job/15764998643

Here's one idea for how we can cleanly add this test data to the PR comments.

----

✨ WebPageTest results for privacy.js

<details>
<summary><strong>✅ Test passed for <a href="https://www.webpagetest.org/results.php?test=230809_BiDc7R_EMG">https://www.theverge.com/</a></strong></summary>

```json
{
    "privacy_wording_links": [
        {
            "keywords": [
                "POLICY"
            ],
            "text": "POLICY"
        },
        {
            "keywords": [
                "Policy"
            ],
            "text": "Policy"
        },
        {
            "keywords": [
                "PRIVACY"
            ],
            "text": "PRIVACY NOTICE"
        },
        {
            "keywords": [
                "COOKIE POLICY"
            ],
            "text": "COOKIE POLICY"
        }
    ],
    "iab_tcf_v1": {
        "present": false,
        "data": null,
        "compliant_setup": null
    },
    "iab_tcf_v2": {
        "present": false,
        "data": null,
        "compliant_setup": null
    },
    "iab_usp": {
        "present": true,
        "privacy_string": {
            "uspString": "1YNY",
            "version": 1
        }
    },
    "ads_transparency_spotlight": {
        "present": false,
        "ads_metadata": null
    },
    "document_interestCohort": false,
    "navigator_doNotTrack": true,
    "navigator_globalPrivacyControl": true,
    "document_permissionsPolicy": false,
    "document_featurePolicy": true,
    "referrerPolicy": {
        "entire_document_policy": null,
        "individual_requests": null,
        "link_relations": {
            "A": 7
        }
    },
    "media_devices": {
        "navigator_mediaDevices_enumerateDevices": false,
        "navigator_mediaDevices_getUserMedia": false,
        "navigator_mediaDevices_getDisplayMedia": false
    },
    "geolocation": {
        "navigator_geolocation_getCurrentPosition": false,
        "navigator_geolocation_watchPosition": false
    }
}
```
</details>

<details>
<summary><strong>✅ Test passed for <a href="https://www.webpagetest.org/results.php?test=230809_BiDcEC_EMK">https://www.rtl.de/</a></strong></summary>

```json
{
    "privacy_wording_links": [
        {
            "keywords": [
                "Datenschutz"
            ],
            "text": "Datenschutz"
        }
    ],
    "iab_tcf_v1": {
        "present": false,
        "data": null,
        "compliant_setup": null
    },
    "iab_tcf_v2": {
        "present": true,
        "data": {
            "cmpId": 6,
            "cmpVersion": 1,
            "gdprApplies": true,
            "tcfPolicyVersion": 2,
            "eventStatus": "cmpuishown",
            "cmpStatus": "loaded",
            "listenerId": null,
            "tcString": "CPwPGYAPwPGYAAGABCENDPCgAAAAAAAAACiQAAAAAAAA.YAAAAAAAAAAA",
            "isServiceSpecific": true,
            "useNonStandardStacks": false,
            "purposeOneTreatment": false,
            "publisherCC": "US",
            "outOfBand": {
                "allowedVendors": {},
                "disclosedVendors": {}
            },
            "purpose": {
                "consents": {},
                "legitimateInterests": {}
            },
            "vendor": {
                "consents": {},
                "legitimateInterests": {}
            },
            "specialFeatureOptins": {},
            "publisher": {
                "consents": {},
                "legitimateInterests": {},
                "customPurpose": {
                    "consents": {},
                    "legitimateInterests": {}
                },
                "restrictions": {}
            },
            "addtlConsent": "1~"
        },
        "compliant_setup": true
    },
    "iab_usp": {
        "present": false,
        "privacy_string": null
    },
    "ads_transparency_spotlight": {
        "present": false,
        "ads_metadata": null
    },
    "document_interestCohort": false,
    "navigator_doNotTrack": true,
    "navigator_globalPrivacyControl": false,
    "document_permissionsPolicy": false,
    "document_featurePolicy": true,
    "referrerPolicy": {
        "entire_document_policy": null,
        "individual_requests": null,
        "link_relations": {
            "A": 35
        }
    },
    "media_devices": {
        "navigator_mediaDevices_enumerateDevices": false,
        "navigator_mediaDevices_getUserMedia": false,
        "navigator_mediaDevices_getDisplayMedia": false
    },
    "geolocation": {
        "navigator_geolocation_getCurrentPosition": true,
        "navigator_geolocation_watchPosition": true
    }
}
```
</details>

----

So we use the `<details>` element to hide the verbose test results, and the `<summary>` element shows the pass/fail status, the name of the test page, and a link to the full test results. Let me know if you think that could be useful.

Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:1]: I've cleaned up logs formatting a bit and added logging of custom metrics to a comment (additional to logs).

Right now I don't see a way to add test results summary there without writing some custom json parser and renderer. I believe somebody should know a more straightforward solution.

**Comment by max-ostapenko** [comment]
As a nice to have added [a failing test annotations](https://jestjs.io/blog/2022/04/25/jest-28#github-actions-reporter).
It's somehow obscure, but may be useful in discussions.

File: tests/privacy.test.js
Comment (rviscomi) [reply tests/privacy.test.js:11]: @tunetheweb any thoughts?

**Comment by tunetheweb** [comment]
Apologies for the radio silence from me. Been travelling a lot and not had time to review this until now.

First up, there's a lot of work here so thanks @max-ostapenko for all efforts here. 🙏

However, I'm not sure this is the right approach...

To me there are two things we could do with automated testing:
1. Test new functionality against a given website and provide a link to the WPT for that site for the custom metric file being changed. This might be temporary URL (e.g. a test site) and may not make a good long term test case, but it shows it works.
2. Have "unit tests" (or "integration" tests) to check existing custom metrics still work and a PR hasn't broken them completely.

This PR seems to address 2. but I think the original aim of #6 was more about 1.

And I do worry about the maintainability of the test cases—especially as this grows to cover each custom metric file and as sites change. Currently, if I'm reading the code correctly, this will fail if a custom metric does not have a test case, and only two of them do with this PR ads and privacy), so there's an initial effort for each person adding a small tweak to an existing custom metric if we merge this. We could pre-initialise all the custom metrics with a basic test case, but that potentially makes maintainability worse—particularly if we want meaningful test cases for each one. Or we keep them light and only add fuller coverage for people who want them to really test all their custom metrics.

So, personally, I'd prefer it if we looked at 1. more, but kept most of the work here for 2. but scaled it down somewhat.

For 1. (testing changes) I think we should have a PR template which takes a URL/URLs that is filled in by the person raising the PR, and runs the changed custom metric JavaScript file(s) against that URL/those URLs. Annoyingly, I don't think GitHub supports [fancy PR templates with fields](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms) (that's only for Issues right now AFAIK), so probably would need to regex the line out of a basic PR template and hope people fill them in right, until GitHub supports that. If no website is given we skip this test (maybe with a warning?).

For 2. (unit tests) I think we pare this back quick a bit and not depend on sites that are out of our control and may change. I would suggest we test httparchive.org (or maybe almanac.httparchive.org?) for ALL custom metrics in one go (like the actual monthly run would do for that site). We do this for every PR (except PRs that just change .md files—like README.md maybe?). This would just be a basic test to ensure all the custom metrics run, and some output is produced. I don't think we need to have checks for each custom metric as many, like ads and privacy may be empty for our sites and that's expected. Maybe we could test a couple of basic tests (e.g. those from markup.js or javascript.js) to ensure some meaningful output is produced for at least some of the custom metrics, but mainly we just want to check that a PR doesn't break everything due to a typo or bad JS.

I know this PR has been open for a while and a lot of work has gone into it, so maybe, as a first step, we could implement one of those suggested workflows in this PR? Then get it merged. Then look at the other one in it's own workflow? I'm open to doing 1 first and then 2 or vice versa.

WDYT @max-ostapenko @rviscomi ?

**Comment by tunetheweb** [comment]
In not sure I understand what "1. in unit test runs skip custom metrics changed (only dist/*.js checked) but without unit tests available" means?

Does this mean "for unit test mode, don't restrict it to the files changed, but always run all the metrics, and as well as just running the custom metrics also run the any unit tests js scripts that exist against the results"?

**Comment by max-ostapenko** [comment]
Clarification on todo 1:
- if dest/ads.js and dest/privacy.js was changed in PR, but no tests/ads.test.js available - we run only tests/privacy.test.js.
- if dest/cms.js was changed in PR, but no tests/cms.test.js available - we don't run any unit tests.

**Comment by rviscomi** [comment]
Thanks @max-ostapenko that plan SGTM.

**Comment by tunetheweb** [comment]
I was thinking more that for all PRs (except maybe simple .md edits):
- all custom metrics run on almanac.httparchive.org. Link to WebPageTest run added to PR.
- all unit tests run on  results of the above run on almanac.httparchive.org
- all custom metrics run on additional URL(s) if provided in PR issue text. Link to WebPageTest run added to PR.

If we only run the unit tests when those custom metrics are changed then it could be years between runs and they might get out of date and stop working if we change almanac.httparchive.org without thinking about this. 

Similarly changing YAML files, or dependencies might break unit tests without us realising.

Running them for every PR means it’s less like to happen.

It’s kinda like we run lots of tests on almanac.httparchive.org PRs even when they don’t touch code that should affect those tests.

**Comment by max-ostapenko** [comment]
You're right, will change the unit test trigger to run on any code changes.

**Comment by rviscomi** [comment]
@max-ostapenko @pmeenan BTW we should also make sure this works with our private instance. At the moment Max is using my personal WPT API key for testing 😅

**Comment by max-ostapenko** [comment]
@tunetheweb @rviscomi 
So:
- We're running unit tests on every PR. WPT link to test run data in comment.
- In PR body you can define list of websites to test additionally (no unit tests run for these). The link + custom metrics values (if changed in PR) available in comment.

File: tests/wpt.js
Comment (max-ostapenko) [inline tests/wpt.js:8]: @rviscomi @pmeenan
We need to change a server and key (value in Github secrets) to start using httparchive private instance.

Comment (max-ostapenko) [reply tests/wpt.js:8]: @pmeenan And I would appreciate a general review of this wpt script to avoid discrepancies with the monthly crawler configuration.

File: package.json
Comment (max-ostapenko) [inline package.json:25]: @rviscomi @tunetheweb could you please fork [catchpoint/WebPageTest.api-nodejs](https://github.com/catchpoint/WebPageTest.api-nodejs) under [HTTPArchive](https://github.com/HTTPArchive) so that the dependency is controlled by the org.

Comment (rviscomi) [reply package.json:25]: Done: https://github.com/HTTPArchive/WebPageTest.api-nodejs

**Review by tunetheweb — APPROVED** [review]
LGTM once we get the keys setup.

Thanks SOOO much for working on this.

**Comment by pmeenan** [comment]
I have an API key allocated for this action for use with the instance at webpagetest.httparchive.org.  I currently have it assigned to the name HA_API_KEY but I can update the existing WPT_API_KEY secret if that would work better. Just need to change the action to point to the other server (or to read the server from a config).

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [inline .github/workflows/wpt-test.yml:31]: @pmeenan with new server we get an error:
```
console.error
      WPTAPIError {
        name: 'WPTAPIError',
        code: 301,
        message: 'Moved Permanently'
      }
```
Please adjust the configuration here.

File: .github/workflows/wpt-test.yml
Comment (max-ostapenko) [inline .github/workflows/wpt-test.yml:32]: @pmeenan with new server we get an error:

console.error
      WPTAPIError {
        name: 'WPTAPIError',
        code: 301,
        message: 'Moved Permanently'
      }
Please adjust the configuration here.

Comment (pmeenan) [reply .github/workflows/wpt-test.yml:32]: Sorry, looks like the client prefers http over https and the server was set to redirect to https.  I've updated the server to allow connections over http as well.

Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:32]: @pmeenan Server responds with 400 now:
```
"statusCode": 400,
"statusText": "An error occurred processing your request (missing API key). If you do not have an API key you can purchase one here: https://product.webpagetest.org/api",
```
Please check the secret.

To test it:
```bash
export WPT_SERVER=webpagetest.httparchive.org
export WPT_API_KEY=INSERT_API_KEY_HERE
npm test
```

Comment (pmeenan) [reply .github/workflows/wpt-test.yml:32]: Sorry about the delay. Not sure why I'm not getting notifications but I'll try to check regularly until it lands.

Not 100% sure why, but it looks like if I pass the key explicitly as part of the options then it works. I submitted the change (hopefully that's ok).

Comment (max-ostapenko) [reply .github/workflows/wpt-test.yml:32]: @pmeenan Thanks for a hint. Looks OK indeed.

**Comment by max-ostapenko** [comment]
@rviscomi no open discussions so far. Ready to merge?

**Comment by rviscomi** [comment]
LGTM! Thank you so much @max-ostapenko!

**Review by rviscomi — APPROVED** [review]

**Comment by tunetheweb** [comment]
Awesome job @max-ostapenko!

**Comment by rviscomi** [comment]
Looks like this failed on a recent PR: https://github.com/HTTPArchive/custom-metrics/actions/runs/6591822240/job/17911210018?pr=96

```
Errors thrown in tests/unit-tests.test.js
    ● _ads parsing
  
      thrown: Object {
        "statusCode": 400,
        "statusText": "An error occurred processing your request (missing API key). If you do not have an API key you can purchase one here: https://product.webpagetest.org/api",
      }
```

**Comment by max-ostapenko** [comment]
@rviscomi Please confirm that `HA_API_KEY` secret (the WPT token) is available.
In the debug logs [a job gets `null`](https://github.com/HTTPArchive/custom-metrics/actions/runs/6591822240/job/17958441417#step:4:7).

**Comment by rviscomi** [comment]
GitHub doesn't actually show the value of the secret, but it seems like it's still there. @pmeenan could you try reapplying it?

image

image

**Comment by tunetheweb** [comment]
Hmmmm this may be a flaw in our plans!

By default, secrets are deliberately not available to PRs from forks. Otherwise an attacker could just create a script to get the secret and open a PR with that script in it.

So not sure how we can get around this...

**Comment by max-ostapenko** [comment]
Yeah, we need to use `pull_request_target` trigger.
Will open a PR.

One more idea:
We could fetch only particular sources from head branch.
As we read and pass `dist/*.js` sources to WPT without executing them, it may be be safe to run WPT for websites from PR description for everyone.
Let me know if this would be valuable (or point to safety issues).

BTW all first time contributors already require approval to run workflows [by default](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks).

**Comment by tunetheweb** [comment]
> We could fetch only particular sources from head branch.
> As we read and pass `dist/*.js` sources to WPT without executing them, it may be be safe to run WPT for websites from PR description for everyone.

Yeah agree that should be fine. In fact because none of the code is executed locally I think the security risk here is low so think we can checkout all the code from head and no need to limit it. In which case can do it for first run too if using `pull_request_target` I think.

The main risk is overwhelming our WebPageTest instance and/or the sites themselves with multiple tests but given we limit it to one worker (I think - @pmeenan can you confirm?) and that new users will require approval as you see there doesn't seem huge risk there either.

**Comment by max-ostapenko** [comment]
PR opened #97

**Comment by max-ostapenko** [comment]
[Some more workflow errors](https://github.com/HTTPArchive/custom-metrics/actions/runs/6624856217/job/17994661665?pr=96) - need to adjust PR description parsing.

**Comment by tunetheweb** [comment]
Good that it ran though!

**Pull request merged**
